receiver/otlpreceiver/metrics/otlp.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics
+
+import (
+	"context"
+
+	collectormetrics "github.com/open-telemetry/opentelemetry-proto/gen/go/collector/metrics/v1"
+	otlpmetrics "github.com/open-telemetry/opentelemetry-proto/gen/go/metrics/v1"
+
+	"github.com/open-telemetry/opentelemetry-collector/client"
+	"github.com/open-telemetry/opentelemetry-collector/consumer"
+	"github.com/open-telemetry/opentelemetry-collector/consumer/consumerdata"
+	"github.com/open-telemetry/opentelemetry-collector/obsreport"
+	"github.com/open-telemetry/opentelemetry-collector/oterr"
+	"github.com/open-telemetry/opentelemetry-collector/translator/metrics"
+)
+
+// Receiver is the type used to handle metrics from OpenTelemetry exporters.
+type Receiver struct {
+	instanceName string
+	nextConsumer consumer.MetricsConsumer
+}
+
+// New creates a new Receiver reference.
+func New(instanceName string, nextConsumer consumer.MetricsConsumer) (*Receiver, error) {
+	if nextConsumer == nil {
+		return nil, oterr.ErrNilNextConsumer
+	}
+	r := &Receiver{
+		instanceName: instanceName,
+		nextConsumer: nextConsumer,
+	}
+	return r, nil
+}
+
+const (
+	receiverTagValue  = "otlp_metrics"
+	receiverTransport = "grpc"
+)
+
+func (r *Receiver) Export(ctx context.Context, req *collectormetrics.ExportMetricsServiceRequest) (*collectormetrics.ExportMetricsServiceResponse, error) {
+	// The bundler will receive batches of metrics i.e. []*otlpmetrics.Metric
+	receiverCtx := obsreport.ReceiverContext(ctx, r.instanceName, receiverTransport, receiverTagValue)
+
+	for _, resourceMetrics := range req.ResourceMetrics {
+		err := r.processReceivedMetrics(receiverCtx, resourceMetrics)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &collectormetrics.ExportMetricsServiceResponse{}, nil
+}
+
+func (r *Receiver) processReceivedMetrics(receiverCtx context.Context, resourceMetrics *otlpmetrics.ResourceMetrics) error {
+	if len(resourceMetrics.InstrumentationLibraryMetrics) == 0 {
+		return nil
```

Comment:
Is this desirable? It will result in no observability data. @pjanotti what do you think? Should we emit own observability if we process empty input data?

Reply:
Keeping until you and Paulo agree on the desired behavior :)

Reply:
Thanks for flagging it @tigrannajaryan: this is not consistent between receivers and even between OC trace and metrics. The better behavior will be to record the observability data but not call the next consumer since there is no data. In a few hours, I will create a PR to make OC consistent and following the behavior described above.

Reply:
Notice that the behavior here is slightly different than OC because it is unary and also the resource metrics end up sent to the next consumer in multiple calls.

Reply:
https://github.com/open-telemetry/opentelemetry-collector/pull/659